mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp

```diff
@@ -1266,6 +1266,121 @@ struct TosaFoldConstantMaximum
   }
 };
 
+template <typename BaseType>
+DenseElementsAttr padType(ShapedType inputType, ElementsAttr inputValues,
+                          DenseElementsAttr paddings,
+                          std::optional<DenseElementsAttr> padConstValue,
+                          ShapedType outputType, BaseType zero) {
+  BaseType padConst(zero);
+  if (padConstValue.has_value())
+    padConst = padConstValue.value().getSplatValue<BaseType>();
+
+  auto values = inputValues.getValues<BaseType>();
+  auto paddingVals = paddings.getValues<int64_t>();
+
+  auto outputShape = outputType.getShape();
+  auto inputShape = inputType.getShape();
+
+  // Implements the logic from
+  // https://www.mlplatform.org/tosa/tosa_spec.html#_pad
+  SmallVector<BaseType> outputValues(outputType.getNumElements(), padConst);
+  for (size_t outIndex = 0, e = outputValues.size(); outIndex < e; ++outIndex) {
+    auto indexInTarget = offsetToIndex(outputShape, outIndex);
+
+    bool isPad =
+        llvm::any_of(llvm::enumerate(indexInTarget), [&](const auto &dimInfo) {
+          auto index = dimInfo.index();
+          auto i = dimInfo.value() - paddingVals[index * 2];
+          return static_cast<bool>(i < 0 || i >= inputShape[index]);
+        });
+
+    auto inputIndexOffset = indexToOffset(outputShape, indexInTarget);
+    outputValues[outIndex] = isPad ? padConst : values[inputIndexOffset];
+  }
+  return DenseElementsAttr::get(outputType,
+                                llvm::ArrayRef<BaseType>(outputValues));
+}
+
+DenseElementsAttr pad(ShapedType inputType, ElementsAttr inputValues,
+                      DenseElementsAttr paddings,
+                      std::optional<DenseElementsAttr> padConstValue,
+                      ShapedType outputType) {
+
+  auto baseType = inputType.getElementType();
+
+  // Handle possible integer types
+  if (auto intType = dyn_cast<IntegerType>(baseType)) {
+    switch (intType.getWidth()) {
+    case 1:
+      return padType<bool>(inputType, inputValues, paddings, padConstValue,
+                           outputType, false);
+    case 8:
+      return padType<int8_t>(inputType, inputValues, paddings, padConstValue,
+                             outputType, 0);
+    case 16:
+      return padType<int16_t>(inputType, inputValues, paddings, padConstValue,
+                              outputType, 0);
+    case 32:
+      return padType<int32_t>(inputType, inputValues, paddings, padConstValue,
+                              outputType, 0);
+    case 64:
```

Comment:
The PAD spec states only up to signed 32. Do we need it then?

Reply:
Simplified to only use `APInt` and `APFloat`. Verifiers should be the ones checking for types.